### ISSUE TITLE
Line terminator problems on Windows

### ISSUE THREAD

**Issue by JFormDesigner** [TomasMikula/RichTextFX#183]
Hi Thomas,

The commit b03e1cc2971884a42ee9ff32d67ac0a2325574f9 causes some new problems with \r\n handling on Windows.

a) Pasting some multiline text (from notepad.exe) throws following exception **when pasting it twice**:
```
C:\Users\charly\Downloads>"C:\Program Files\Java\jdk1.8.0_60\bin\java" -cp richtextfx-demos-fat-0.6.9.jar org.fxmisc.richtext.demo.JavaKeywords
Exception in thread "JavaFX Application Thread" java.lang.StringIndexOutOfBoundsException: String index out of range: -2
        at java.lang.String.substring(String.java:1967)
        at org.fxmisc.richtext.Paragraph.substring(Paragraph.java:73)
        at org.fxmisc.richtext.StyledDocumentBase.sub(StyledDocumentBase.java:245)
        at org.fxmisc.richtext.StyledDocumentBase.getText(StyledDocumentBase.java:51)
        at org.fxmisc.richtext.EditableStyledDocument.lambda$null$22(EditableStyledDocument.java:129)
        at org.reactfx.util.Tuple2.map(Tuple2.java:31)
        at org.fxmisc.richtext.EditableStyledDocument.lambda$new$23(EditableStyledDocument.java:129)
        at org.reactfx.MappedStream.lambda$observeInputs$201(MappedStream.java:22)
        at org.reactfx.util.NonAccumulativeStreamNotifications.lambda$head$272(NotificationAccumulator.java:134)
        at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
        at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
        at org.reactfx.ProperEventStream.emit(ProperEventStream.java:18)
        at org.reactfx.EventStreams$20.tryEmit(EventStreams.java:441)
        at org.reactfx.EventStreams$20.lambda$observeInputs$36(EventStreams.java:436)
        at org.reactfx.util.NonAccumulativeStreamNotifications.lambda$head$272(NotificationAccumulator.java:134)
        at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
        at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:57)
        at org.reactfx.ProperEventStream.emit(ProperEventStream.java:18)
        at org.reactfx.EventSource.push(EventSource.java:18)
        at org.fxmisc.richtext.EditableStyledDocument.replace(EditableStyledDocument.java:454)
        at org.fxmisc.richtext.EditableStyledDocument.replaceText(EditableStyledDocument.java:192)
        at org.fxmisc.richtext.StyledTextArea.replaceText(StyledTextArea.java:720)
        at org.fxmisc.richtext.TextEditingArea.replaceText(TextEditingArea.java:204)
        at org.fxmisc.richtext.EditActions.replaceSelection(EditActions.java:119)
        at org.fxmisc.richtext.ClipboardActions.paste(ClipboardActions.java:97)
        at org.fxmisc.richtext.skin.StyledTextAreaBehavior.lambda$static$69(StyledTextAreaBehavior.java:65)
        at org.fxmisc.wellbehaved.event.EventHandlerTemplate$On.lambda$null$23(EventHandlerTemplate.java:156)
        at java.util.Optional.ifPresent(Optional.java:159)
        at org.fxmisc.wellbehaved.event.EventHandlerTemplate$On.lambda$act$24(EventHandlerTemplate.java:155)
        at org.fxmisc.wellbehaved.event.EventHandlerTemplate$Builder.lambda$create$21(EventHandlerTemplate.java:117)
        at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$17(EventHandlerTemplate.java:39)
        at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$15(EventHandlerTemplate.java:26)
        at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$19(EventHandlerTemplate.java:49)
        at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$15(EventHandlerTemplate.java:26)
        at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$15(EventHandlerTemplate.java:26)
        at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$15(EventHandlerTemplate.java:26)
        at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$bind$14(EventHandlerTemplate.java:18)
        at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
        at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
        at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
        at javafx.event.Event.fireEvent(Event.java:198)
        at javafx.scene.Scene$KeyHandler.process(Scene.java:3964)
        at javafx.scene.Scene$KeyHandler.access$1800(Scene.java:3910)
        at javafx.scene.Scene.impl_processKeyEvent(Scene.java:2040)
        at javafx.scene.Scene$ScenePeerListener.keyEvent(Scene.java:2501)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:197)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:147)
        at java.security.AccessController.doPrivileged(Native Method)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleKeyEvent$354(GlassViewEventHandler.java:228)
        at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleKeyEvent(GlassViewEventHandler.java:227)
        at com.sun.glass.ui.View.handleKeyEvent(View.java:546)
        at com.sun.glass.ui.View.notifyKey(View.java:966)
        at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at com.sun.glass.ui.win.WinApplication.lambda$null$149(WinApplication.java:191)
        at java.lang.Thread.run(Thread.java:745)
```

b) `EditableStyledDocument.getLength()` returns a too large length (including \r characters)
```java
@Test
public void testWinDocumentLength() {
    EditableStyledDocument<String> document = new EditableStyledDocument<>("");
    document.replaceText(0, 0, "X\r\nY");
    assertEquals(document.getText().length(), document.getLength());
}
```
throws `java.lang.AssertionError: expected:<3> but was:<4>`

c) after pasting a text that contains \r characters, the caret is not at the end of the pasted text. It is some characters behind it, which depends on the number of \r characters in the pasted text.

d) Undo after pasting a text that contains \r characters removes too many text.

Looking at the source code, one reason may be the usage of `text.length()` where `text` contains \r characters, but paragraphs no longer contain them. There several places. E.g.

* https://github.com/TomasMikula/RichTextFX/blob/master/richtextfx/src/main/java/org/fxmisc/richtext/EditableStyledDocument.java#L473
* https://github.com/TomasMikula/RichTextFX/blob/master/richtextfx/src/main/java/org/fxmisc/richtext/EditableStyledDocument.java#L196
* https://github.com/TomasMikula/RichTextFX/blob/master/richtextfx/src/main/java/org/fxmisc/richtext/StyledTextArea.java#L722

Don't know how to fix this best. Maybe always do `text.replace("\r\n", "\n")` on all inserted strings.

Karl

**Comment by TomasMikula** [comment]
Thanks for providing a test case. I fixed your point b) in a recent commit. Not sure how it affects the other points. Would be great if you could try it out. Obviously, there's no release with this fix yet, but I updated the [1.0.0-SNAPSHOT](https://github.com/TomasMikula/RichTextFX#snapshot-releases), if you want to give it a try.

**Comment by JFormDesigner** [comment]
Thanks. b) and also a) are gone.

c) and d) are still there.

c) should be easy to fix by changing [StyledTextArea.java#L722](https://github.com/TomasMikula/RichTextFX/blob/master/richtextfx/src/main/java/org/fxmisc/richtext/StyledTextArea.java#L722) to
`int newCaretPos = start + text.replace("\r\n", "\n").length();`

Here is a little App to test c) and d)
```java
public class RichTextWinUndoTest extends Application {
	@Override
	public void start(Stage primaryStage) {
		String unixText1 = "abc\ndef\nghi";
		String unixText2 = "A\nB\nC";
		String winText1 = unixText1.replace("\n", "\r\n");
		String winText2 = unixText2.replace("\n", "\r\n");

		StyleClassedTextArea winTextArea = new StyleClassedTextArea();
		winTextArea.replaceText(winText1);
		winTextArea.getUndoManager().forgetHistory();
		winTextArea.insertText(0,winText2);

		StyleClassedTextArea unixTextArea = new StyleClassedTextArea();
		unixTextArea.replaceText(unixText1);
    	unixTextArea.getUndoManager().forgetHistory();
		unixTextArea.insertText(0,unixText2);

		Scene scene = new Scene(new SplitPane(winTextArea, unixTextArea), 600, 400);
		primaryStage.setScene(scene);
		primaryStage.setTitle("RichTextFX");
		primaryStage.show();
	}
}
```
After starting the App, the caret in the left (Windows) text area is after 'b', but should after 'C'.

Pressing Ctrl+Z in the left (Windows) text area removes not only the uppercase characters, but also the lowercase "ab". Pressing Ctrl+Z in the right (Unix) text area removes only the uppercase characters.

When playing arround with pasting multiline text (with \r\n copied from notepad.exe) and undo, sometimes I get following exception (don't have a test case for this):
```
Exception in thread "JavaFX Application Thread" java.lang.IllegalArgumentException: Unexpected change received.
Expected:
org.fxmisc.richtext.RichTextChange@d06fec6
Received:
org.fxmisc.richtext.RichTextChange@b5e6df07
	at org.fxmisc.undo.impl.UndoManagerImpl.changeObserved(UndoManagerImpl.java:185)
	at org.reactfx.util.QueuingStreamNotifications.lambda$head$274(NotificationAccumulator.java:217)
	at org.reactfx.ObservableBase.notifyObservers(ObservableBase.java:68)
	at org.reactfx.SuspendableBase.resume(SuspendableBase.java:64)
	at org.reactfx.CloseableOnceGuard.close(Guard.java:49)
	at org.reactfx.MultiGuard.close(Guard.java:83)
	at org.fxmisc.richtext.StyledTextArea.replace(StyledTextArea.java:737)
	at org.fxmisc.richtext.StyledTextArea.lambda$createRichUndoManager$11(StyledTextArea.java:799)
	at org.fxmisc.undo.impl.UndoManagerImpl.lambda$performChange$5(UndoManagerImpl.java:176)
	at org.reactfx.Suspendable.suspendWhile(Suspendable.java:47)
	at org.fxmisc.undo.impl.UndoManagerImpl.performChange(UndoManagerImpl.java:176)
	at org.fxmisc.undo.impl.UndoManagerImpl.undo(UndoManagerImpl.java:94)
	at org.fxmisc.richtext.UndoActions.undo(UndoActions.java:21)
	at org.fxmisc.richtext.skin.StyledTextAreaBehavior.lambda$static$73(StyledTextAreaBehavior.java:71)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate$On.lambda$null$23(EventHandlerTemplate.java:156)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate$On.lambda$act$24(EventHandlerTemplate.java:155)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate$Builder.lambda$create$21(EventHandlerTemplate.java:117)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$17(EventHandlerTemplate.java:39)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$15(EventHandlerTemplate.java:26)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$19(EventHandlerTemplate.java:49)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$15(EventHandlerTemplate.java:26)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$15(EventHandlerTemplate.java:26)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$null$15(EventHandlerTemplate.java:26)
	at org.fxmisc.wellbehaved.event.EventHandlerTemplate.lambda$bind$14(EventHandlerTemplate.java:18)
	at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.event.Event.fireEvent(Event.java:198)
	at javafx.scene.Scene$KeyHandler.process(Scene.java:3964)
	at javafx.scene.Scene$KeyHandler.access$1800(Scene.java:3910)
	at javafx.scene.Scene.impl_processKeyEvent(Scene.java:2040)
	at javafx.scene.Scene$ScenePeerListener.keyEvent(Scene.java:2501)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:197)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:147)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleKeyEvent$354(GlassViewEventHandler.java:228)
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleKeyEvent(GlassViewEventHandler.java:227)
	at com.sun.glass.ui.View.handleKeyEvent(View.java:546)
	at com.sun.glass.ui.View.notifyKey(View.java:966)
	at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at com.sun.glass.ui.win.WinApplication.lambda$null$149(WinApplication.java:191)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by TomasMikula** [comment]
Thanks for helping me hunting it down. It should be fixed now in 1.0.0-SNAPSHOT.

**Comment by JFormDesigner** [comment]
Yes, it works now. Many thanks.

**Comment by TomasMikula** [comment]
I just published `0.6.10` with this fix.
